File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    vector<string>res;
    void help(int n , string path , int open , int close){
        if(n == 0 && open==close){
            res.push_back(path);
            return;
        }
        if(n==0){
            return;
        }
        if(open>close){
            help(n-1,path+'(',open+1,close);
            help(n-1,path+')',open,close+1);
        }else{
            help(n-1,path+'(',open+1,close);
        }
    }
    vector<string> generateParenthesis(int n) {
        n = 2* n;
        help(n,"",0,0);
        return res;
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    
    List<String> res = new ArrayList<>();
    
    
    public void help(int open , int close ,String path, int n){
        
        if(close == n && open == n){
            res.add(path);
            return ;
        }
        
        if(open > n || (close == n && open != n)){
            return ;
        }
        
        if(open >= close){
            
            help(open+1,close,path+'(',n);
            help(open,close+1,path+')',n);
        }
        else{
            help(open,close+1,path+')',n);
        }
        
    }
    
    public List<String> generateParenthesis(int n) {
        String path = new String();
        int open = 0;
        int close = 0;
        help(open,close,path,n);
        return res;
    }
}